torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -756,6 +769,38 @@ void Reducer::finalize_backward() {
     local_used_work_->wait();
   }
   local_used_maps_reduced_ = false;
+  rpc_context_.clear();
```

Comment:
I think we should get rid of the `rpc_context_` and use the suggestion of `auto context_ptr = ThreadLocalDistAutogradContext::getContextPtr()` mentioned below. One of the reasons is that it is pretty tricky to ensure `rpc_context_` is cleared correctly. The code before this could throw an exception and we wouldn't clear rpc_context_.

The callbacks are executed in the same thread post hook and as a result the thread local should be valid at that time. I double checked that all `runGradCallbackForVariable` would be executed in the same thread as the post hook and as a result they would pick up the thread local. The only exception here is `finalize_backward` and as a result we should set the thread local for that call as well. I feel this would be much cleaner since we don't need to worry about the lifetime of rpc_context_ and clear it at the appropriate time.

Reply:
@pritamdamania87 

I don't have a strong opinion here,  and just prefer another structure. To ensure `rpc_context_` is cleared correctly, we can use a guard here to ensure that.
The delayed thread local rpc context retrieval works now, but it would be quite easy to break in the future. Besides that, to set thread local for the call of `finalize_backward`, does `Reduce` need to hold the rpc context anyway to do that?
What do you think?

Reply:
> I don't have a strong opinion here, and just prefer another structure. To ensure rpc_context_ is cleared correctly, we can use a guard here to ensure that.

This doesn't really solve the problem, there could be an exception in the autograd engine and finalize_backward might not be called at all.

> Besides that, to set thread local for the call of finalize_backward, does Reduce need to hold the rpc context anyway to do that?

This line is called in the post hook and the thread local should be set there if we're using distributed autograd. https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L394. You can check for this before that line and if the thread local is set, you can capture the thread local in the lambda and ensure it is set in the lambda function before `finalize_backward` is called.

Reply:
Resolved offline